runtime/src/main/cpp/Memory.h

```diff
@@ -38,19 +38,29 @@ struct MetaObjHeader;
 struct ObjHeader {
   TypeInfo* typeInfoOrMeta_;
 
+  // Returns `nullptr` if it's not a meta object.
+  static MetaObjHeader* AsMetaObject(TypeInfo* typeInfo) noexcept {
+      auto* typeInfoOrMeta = clearPointerBits(typeInfo, OBJECT_TAG_MASK);
+      if (typeInfoOrMeta != typeInfoOrMeta->typeInfo_) {
+          return reinterpret_cast<MetaObjHeader*>(typeInfoOrMeta);
+      } else {
+          return nullptr;
+      }
+  }
+
   const TypeInfo* type_info() const {
     return clearPointerBits(typeInfoOrMeta_, OBJECT_TAG_MASK)->typeInfo_;
   }
 
   bool has_meta_object() const {
-    auto* typeInfoOrMeta = clearPointerBits(typeInfoOrMeta_, OBJECT_TAG_MASK);
-    return (typeInfoOrMeta != typeInfoOrMeta->typeInfo_);
+      return AsMetaObject(typeInfoOrMeta_) != nullptr;
```

Comment:
This might be suboptimal: does the compiler optimize this code by eliminating the redundant conditional jump here?
I mean, if `typeInfoOrMeta != typeInfoOrMeta->typeInfo_` does the compiler assume that `typeInfoOrMeta != nullptr` to avoid checking this explicitly here?

Reply:
Looking at the code:
```cpp
    if (!has_meta_object()) {
        return nullptr;
    }
    return (void*)1;
```
With the old `has_meta_object`:
```llvm
  %2 = bitcast %struct.ObjHeader* %0 to i64*
  %3 = load i64, i64* %2, align 8, !tbaa !4
  %4 = and i64 %3, -4
  %5 = inttoptr i64 %4 to %struct.TypeInfo*
  %6 = getelementptr inbounds %struct.TypeInfo, %struct.TypeInfo* %5, i64 0, i32 0
  %7 = load %struct.TypeInfo*, %struct.TypeInfo** %6, align 8, !tbaa !9
  %8 = icmp eq %struct.TypeInfo* %7, %5
  %9 = select i1 %8, i8* null, i8* inttoptr (i64 1 to i8*)
  ret i8* %9
```
With the new:
```llvm
  %2 = bitcast %struct.ObjHeader* %0 to i64*
  %3 = load i64, i64* %2, align 8, !tbaa !4
  %4 = and i64 %3, -4
  %5 = inttoptr i64 %4 to %struct.TypeInfo*
  %6 = getelementptr inbounds %struct.TypeInfo, %struct.TypeInfo* %5, i64 0, i32 0
  %7 = load %struct.TypeInfo*, %struct.TypeInfo** %6, align 8, !tbaa !9
  %8 = icmp ne %struct.TypeInfo* %7, %5
  %9 = icmp ne i64 %4, 0
  %10 = and i1 %9, %8
  %11 = select i1 %10, i8* inttoptr (i64 1 to i8*), i8* null
  ret i8* %11
}
```
So, it doesn't generate an extra jump, but does generate an extra `and`. Same in the x64 assembly.

Reply:
Also an extra comparison. But this might be ok.